File: solutions/typescript/128-graph-valid-tree.ts
/**
 * Problem 128: Graph Valid Tree (LeetCode 261)
 * Difficulty: Med
 * Language: TypeScript
 */
function validTree(n: number, edges: number[][]): boolean {
  if (edges.length !== n - 1) return false;
  const adj = edges.reduce<Record<number, number[]>>(
    (g, [u, v]) => ({
      ...g, [u]: [...(g[u] || []), v], [v]: [...(g[v] || []), u]
    }), {});
  const dfs = (node: number, visited: Set<number>): Set<number> => {
    if (visited.has(node)) return visited;
    const v = new Set([...visited, node]);
    return (adj[node] || []).reduce((acc, nb) => dfs(nb, acc), v);
  };
  return dfs(0, new Set()).size === n;
}

File: solutions/elixir/128-graph-valid-tree.ex
# Problem 128: Graph Valid Tree (LeetCode 261)
# Difficulty: Med
# Language: Elixir
# 
defmodule ValidTree do
  def valid_tree(n, edges) do
    if length(edges) != n - 1, do: false,
    else: (
      adj = Enum.reduce(edges, %{}, fn [u, v], g ->
        g |> Map.update(u, [v], &[v | &1])
          |> Map.update(v, [u], &[u | &1])
      end)
      visited = dfs(0, MapSet.new(), adj)
      MapSet.size(visited) == n
    )
  end
  defp dfs(node, visited, adj) do
    if MapSet.member?(visited, node), do: visited,
    else: (
      visited = MapSet.put(visited, node)
      Map.get(adj, node, [])
      |> Enum.reduce(visited, fn nb, acc -> dfs(nb, acc, adj) end)
    )
  end
end
